extensions/aws/processors/PutS3Object.cpp

```diff
@@ -0,0 +1,487 @@
+  context->getProperty(ProxyPassword, proxy.password, flow_file);
+  if (!proxy.host.empty()) {
+    s3_wrapper_->setProxy(proxy);
+    logger_->log_info("Proxy for PutS3Object was set.");
+  }
+  return true;
+}
+
+void PutS3Object::onSchedule(const std::shared_ptr<core::ProcessContext> &context, const std::shared_ptr<core::ProcessSessionFactory> &sessionFactory) {
+  if (!context->getProperty(Bucket.getName(), put_s3_request_params_.bucket) || put_s3_request_params_.bucket.empty()) {
+    throw Exception(PROCESS_SCHEDULE_EXCEPTION, "Bucket property missing or invalid");
+  }
+  logger_->log_debug("PutS3Object: Bucket [%s]", put_s3_request_params_.bucket);
+
+  if (!context->getProperty(StorageClass.getName(), put_s3_request_params_.storage_class)
+      || put_s3_request_params_.storage_class.empty()
+      || STORAGE_CLASSES.find(put_s3_request_params_.storage_class) == STORAGE_CLASSES.end()) {
+    throw Exception(PROCESS_SCHEDULE_EXCEPTION, "Storage Class property missing or invalid");
+  }
+  logger_->log_debug("PutS3Object: Storage Class [%s]", put_s3_request_params_.storage_class);
+
+  std::string value;
+  if (!context->getProperty(Region.getName(), value) || value.empty() || REGIONS.find(value) == REGIONS.end()) {
+    throw Exception(PROCESS_SCHEDULE_EXCEPTION, "Region property missing or invalid");
+  }
+  s3_wrapper_->setRegion(value);
+  logger_->log_debug("PutS3Object: Region [%s]", value);
+
+  uint64_t timeout_val;
+  if (context->getProperty(CommunicationsTimeout.getName(), value) && !value.empty() && core::Property::getTimeMSFromString(value, timeout_val)) {
+    s3_wrapper_->setTimeout(timeout_val);
+    logger_->log_debug("PutS3Object: Communications Timeout [%d]", timeout_val);
+  } else {
+    throw Exception(PROCESS_SCHEDULE_EXCEPTION, "Communications Timeout missing or invalid");
+  }
+
+  if (!context->getProperty(ServerSideEncryption.getName(), put_s3_request_params_.server_side_encryption)
+      || put_s3_request_params_.server_side_encryption.empty()
+      || SERVER_SIDE_ENCRYPTIONS.find(put_s3_request_params_.server_side_encryption) == SERVER_SIDE_ENCRYPTIONS.end()) {
+    throw Exception(PROCESS_SCHEDULE_EXCEPTION, "Server Side Encryption property missing or invalid");
+  }
+  logger_->log_debug("PutS3Object: Server Side Encryption [%s]", put_s3_request_params_.server_side_encryption);
+
+  fillUserMetadata(context);
+}
+
+std::string PutS3Object::parseAccessControlList(const std::string &comma_separated_list) const {
+  std::string result_list;
+  bool is_first = true;
+  for (const auto& user : minifi::utils::StringUtils::split(comma_separated_list, ",")) {
+    if (is_first) {
+      is_first = false;
+    } else {
+      result_list += ", ";
+    }
+
+    auto trimmed_user = minifi::utils::StringUtils::trim(user);
+    static const std::regex email_pattern("(\\w+)(\\.|_)?(\\w*)@(\\w+)(\\.(\\w+))+");
+    if (std::regex_match(trimmed_user, email_pattern)) {
```

Comment:
doesn't `\w` also match `_`?
also note that this won't match all valid email addresses (if that is the purpose), the regex for that is quite... a mouthful

Reply:
Yes, you are right and it is unnecessary to complicate it in this case.

Reply:
Fixed in [2e71a70](https://github.com/apache/nifi-minifi-cpp/pull/915/commits/2e71a707a31f5564360c40c7bfdd38c3d97302ef)

Reply:
the regex rejects a number of otherwise valid email addresses, e.g. the "local-part" (part before the '@' character) may contain "!~#" and others, it could even be quoted or contain comments, the domain part is similarly complex
I don't know what the requirement is here, do we aim to catch all valid email addresses? (if not, could checking for the '@' character be enough?)

Reply:
The goal here would be to distinguish email addresses and AWS ids . I am not sure if quotes and comments should be accepted in this case, but for our goal I suppose checking the '@' character should be enough.

Reply:
is the format of AWS id specified? or is it "not email address"?

Reply:
In this case as we only accept email addresses and AWS ids so currently I only checked if it is an email address or not in which case we assume it is an AWS id. The canonical user id is not really specified it is only defined [here](https://docs.aws.amazon.com/general/latest/gr/acct-identifiers.html) as "An alpha-numeric identifier, such as 79a59df900b949e55d96a1e698fbacedfd6e09d98eacf8f8d5218e7cd47ef2be, that is an obfuscated form of the AWS account ID.". It seems to me it is a 64 long hex string literal, but I am not sure if the length is fixed as there are some questions about it [here for example](https://forums.aws.amazon.com/thread.jspa?threadID=286019).

Reply:
Checking only for '@' for email addresses in [967b355](https://github.com/apache/nifi-minifi-cpp/pull/915/commits/967b355ca776f8a2dba705e390a91bc6aa5380a3)